common/src/main/java/io/netty/util/internal/StringUtil.java

```diff
@@ -260,67 +260,80 @@ public static CharSequence escapeCsv(CharSequence value) {
-        boolean quoted = isDoubleQuote(value.charAt(start)) && isDoubleQuote(value.charAt(last)) && length != 1;
-        boolean foundSpecialCharacter = false;
-        boolean escapedDoubleQuote = false;
-        for (int i = start; i <= last; i++) {
-            char current = value.charAt(i);
-            switch (current) {
-                case DOUBLE_QUOTE:
-                    if (i == start || i == last) {
-                        if (!quoted) {
-                            result.append(DOUBLE_QUOTE);
-                        } else {
-                            continue;
-                        }
-                    } else {
-                        boolean isNextCharDoubleQuote = isDoubleQuote(value.charAt(i + 1));
-                        if (!isDoubleQuote(value.charAt(i - 1)) &&
-                                (!isNextCharDoubleQuote || i + 1 == last)) {
-                            result.append(DOUBLE_QUOTE);
-                            escapedDoubleQuote = true;
+        if (firstUnescapedSpecial < 0) {
+            if (quoted) {
+                for (int i = start; i <= last; i++) {
+                    if (isDoubleQuote(value.charAt(i))) {
+                        if (i == last || !isDoubleQuote(value.charAt(i + 1))) {
+                            firstUnescapedSpecial = i;
+                            break;
                         }
+                        i++;
+                    }
+                }
+            } else {
+                for (int i = start; i <= last; i++) {
+                    char c = value.charAt(i);
+                    if (c == LINE_FEED || c == CARRIAGE_RETURN || c == COMMA) {
+                        firstUnescapedSpecial = i;
                         break;
                     }
-                case LINE_FEED:
-                case CARRIAGE_RETURN:
-                case COMMA:
-                    foundSpecialCharacter = true;
+                    if (isDoubleQuote(c)) {
+                        if (i == last || !isDoubleQuote(value.charAt(i + 1))) {
+                            firstUnescapedSpecial = i;
+                            break;
+                        }
+                        i++;
+                    }
+                }
             }
-            result.append(current);
-        }
 
-        if (escapedDoubleQuote || foundSpecialCharacter && !quoted) {
-            return quote(result);
-        }
-        if (trimmed) {
-            return quoted ? quote(result) : result;
+            if (firstUnescapedSpecial < 0) {
+                // Special characters is not found or all of them already escaped.
+                // In the most cases returns a same string. New string will be instantiated (via StringBuilder)
+                // only if it really needed. It's important to prevent GC extra load.
+                return quoted? value.subSequence(start - 1, last + 2) : value.subSequence(start, last + 1);
```

Comment:
`last + 2` passed if `quoted` is true. The test `StringUtilTest#escapeCsvUnquoted` has no quotes -
 `quoted` is false. The `+2` used because `last` is decremented above:
```java
            if (quoted) {
                start++;
                last--;
            } else {
```